### PR TITLE
set clickhouse max connections

### PR DESCRIPTION
**Pull request by pyama86**
ref: https://github.com/getsentry/snuba/pull/4988
ref: https://github.com/sentry-kubernetes/charts/issues/977

Thank you always for the maintenance.
I've written a patch to fix the error Connection pool is full, discarding connection: sentry-clickhouse. Connection pool size: 1, so please check it.

### CODE REVIEW

**Comment by begemotik** [comment]
To actually merge the PR please bump the Chart version as well as ping the chart maintainer. It will increase the chances that it is merged :)

**Comment by pyama86** [comment]
@Mokto could you check this PR?

File: sentry/templates/configmap-snuba.yaml [hunk 1]
```diff
@@ -25,6 +25,7 @@ data:
         "port": int({{ include "sentry.clickhouse.port" . }}),
         "user":  env("CLICKHOUSE_USER", "default"),
         "password": env("CLICKHOUSE_PASSWORD", ""),
+        "max_connections": int(os.environ.get("CLICKHOUSE_MAX_CONNECTIONS", 100)),
```
Comment (Mokto) [inline sentry/templates/configmap-snuba.yaml:28]: How & when is that env variable set up?

Comment (pyama86) [reply sentry/templates/configmap-snuba.yaml:28]: https://github.com/sentry-kubernetes/charts/pull/1175/commits/8a3b306e87329f57e9e1b4d9270674aad0eebffb

In this commit, we've made it possible to specify from the values.

**Comment by Mokto** [comment]
Perfect, thanks.

**Comment by ajith-nataraj** [comment]
Did this change actually resolve the issue ?
i have changed the CLICKHOUSE_MAX_CONNECTIONS : 100 by env value . still i see the pods getting restarted

**Comment by bkk-bcd** [comment]
> Did this change actually resolve the issue ? i have changed the CLICKHOUSE_MAX_CONNECTIONS : 100 by env value . still i see the pods getting restarted

Ditto:

```
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq 2024-04-17 15:07:28,785 Connection pool is full, discarding connection: sentry-clickhouse. Connection pool size: 1
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq 2024-04-17 15:07:39,113 Caught exception, shutting down...
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq Traceback (most recent call last):
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/processor.py", line 321, in run
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self._run_once()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/processor.py", line 395, in _run_once
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__processing_strategy.poll()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/strategies/guard.py", line 101, in poll
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__inner_strategy.poll()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/strategies/run_task.py", line 55, in poll
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__next_step.poll()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/strategies/guard.py", line 37, in poll
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__next_step.poll()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/strategies/reduce.py", line 101, in poll
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__buffer_step.poll()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/strategies/buffer.py", line 169, in poll
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__next_step.poll()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/strategies/run_task_in_threads.py", line 107, in poll
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     result = future.result()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq              ^^^^^^^^^^^^^^^
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 449, in result
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     return self.__get_result()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq            ^^^^^^^^^^^^^^^^^^^
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     raise self._exception
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 58, in run
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     result = self.fn(*self.args, **self.kwargs)
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/src/snuba/snuba/consumers/strategy_factory.py", line 137, in flush_batch
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     message.payload.close()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/src/snuba/snuba/consumers/consumer.py", line 300, in close
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__insert_batch_writer.close()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/src/snuba/snuba/consumers/consumer.py", line 136, in close
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__writer.write(
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/src/snuba/snuba/clickhouse/http.py", line 359, in write
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     batch.join(timeout=batch_join_timeout)
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/src/snuba/snuba/clickhouse/http.py", line 243, in join
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     response = self._result.result(timeout)
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 458, in result
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     raise TimeoutError()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq TimeoutError
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq 2024-04-17 15:07:39,945 Connection pool is full, discarding connection: sentry-clickhouse. Connection pool size: 1
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq 2024-04-17 15:07:39,972 Connection pool is full, discarding connection: sentry-clickhouse. Connection pool size: 1
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq 2024-04-17 15:07:39,980 Connection pool is full, discarding connection: sentry-clickhouse. Connection pool size: 1
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq 2024-04-17 15:07:39,981 Closing <arroyo.backends.kafka.consumer.KafkaConsumer object at 0x7f65f5022a90>...
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq 2024-04-17 15:07:39,984 Partitions to revoke: [Partition(topic=Topic(name='outcomes'), index=0)]
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq 2024-04-17 15:07:39,984 Partition revocation complete.
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq 2024-04-17 15:07:39,986 Processor terminated
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq Traceback (most recent call last):
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/bin/snuba", line 33, in <module>
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     sys.exit(load_entry_point('snuba', 'console_scripts', 'snuba')())
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     return self.main(*args, **kwargs)
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq            ^^^^^^^^^^^^^^^^^^^^^^^^^^
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1055, in main
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     rv = self.invoke(ctx)
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq          ^^^^^^^^^^^^^^^^
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1657, in invoke
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     return _process_result(sub_ctx.command.invoke(sub_ctx))
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1404, in invoke
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     return ctx.invoke(self.callback, **ctx.params)
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/click/core.py", line 760, in invoke
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     return __callback(*args, **kwargs)
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/src/snuba/snuba/cli/consumer.py", line 284, in consumer
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     consumer.run()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/processor.py", line 321, in run
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self._run_once()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/processor.py", line 395, in _run_once
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__processing_strategy.poll()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/strategies/guard.py", line 101, in poll
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__inner_strategy.poll()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/strategies/run_task.py", line 55, in poll
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__next_step.poll()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/strategies/guard.py", line 37, in poll
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__next_step.poll()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/strategies/reduce.py", line 101, in poll
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__buffer_step.poll()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/strategies/buffer.py", line 169, in poll
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__next_step.poll()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/site-packages/arroyo/processing/strategies/run_task_in_threads.py", line 107, in poll
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     result = future.result()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq              ^^^^^^^^^^^^^^^
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 449, in result
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     return self.__get_result()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq            ^^^^^^^^^^^^^^^^^^^
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     raise self._exception
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 58, in run
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     result = self.fn(*self.args, **self.kwargs)
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/src/snuba/snuba/consumers/strategy_factory.py", line 137, in flush_batch
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     message.payload.close()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/src/snuba/snuba/consumers/consumer.py", line 300, in close
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__insert_batch_writer.close()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/src/snuba/snuba/consumers/consumer.py", line 136, in close
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     self.__writer.write(
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/src/snuba/snuba/clickhouse/http.py", line 359, in write
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     batch.join(timeout=batch_join_timeout)
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/src/snuba/snuba/clickhouse/http.py", line 243, in join
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     response = self._result.result(timeout)
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq   File "/usr/local/lib/python3.11/concurrent/futures/_base.py", line 458, in result
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq     raise TimeoutError()
sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq TimeoutError
Stream closed EOF for sentry/sentry-snuba-outcomes-consumer-54655c6d8b-vfbpq (sentry-snuba)

```

**Comment by jhass** [comment]
Yes, it seems like the fix doesn't actually work. I tried to dig through the snuba code and the PR that introduced the max connections setting and couldn't understand how the setting would actually ever be propagated from the configuration to the initializer arguments, so perhaps it is not?

I'm using helmfile to deploy so for now I'm doing this atrocious hack:

```
releases:
  - name: sentry
    # ...
    transformers:
    - apiVersion: builtin
      kind: PatchJson6902Transformer
      metadata:
        name: patch-sentry-snuba-clickhouse-connections
      target:
        group: apps
        version: v1
        kind: Deployment
        name: sentry-snuba-.*-consumer
      jsonOp: |-
        - op: add
          path: /spec/template/spec/containers/0/command/0
          value: "sh"
        - op: add
          path: /spec/template/spec/containers/0/command/1
          value: "-c"
        - op: add
          path: /spec/template/spec/containers/0/command/2
          value: eval "$@"
        - op: add
          path: /spec/template/spec/containers/0/command/3
          value: "--"
        - op: add
          path: /spec/template/spec/containers/0/command/4
          value: "sed -i 's/max_connections: int = 1/max_connections: int = 100/' snuba/clusters/cluster.py;"
     # ...
```

**Comment by bkk-bcd** [comment]
> Yes, it seems like the fix doesn't actually work. I tried to dig through the snuba code and the PR that introduced the max connections setting and couldn't understand how the setting would actually ever be propagated from the configuration to the initializer arguments, so perhaps it is not?
> 
> I'm using helmfile to deploy so for now I'm doing this atrocious hack:
> 
> ```
> releases:
>   - name: sentry
>     # ...
>     transformers:
>     - apiVersion: builtin
>       kind: PatchJson6902Transformer
>       metadata:
>         name: patch-sentry-snuba-clickhouse-connections
>       target:
>         group: apps
>         version: v1
>         kind: Deployment
>         name: sentry-snuba-.*-consumer
>       jsonOp: |-
>         - op: add
>           path: /spec/template/spec/containers/0/command/0
>           value: "sh"
>         - op: add
>           path: /spec/template/spec/containers/0/command/1
>           value: "-c"
>         - op: add
>           path: /spec/template/spec/containers/0/command/2
>           value: eval "$@"
>         - op: add
>           path: /spec/template/spec/containers/0/command/3
>           value: "--"
>         - op: add
>           path: /spec/template/spec/containers/0/command/4
>           value: "sed -i 's/max_connections: int = 1/max_connections: int = 100/' snuba/clusters/cluster.py;"
>      # ...
> ```

Alternatively:

```Dockerfile
FROM your-base-image

# Copy the custom entrypoint script
COPY custom-entrypoint.sh /custom-entrypoint.sh

# Make the script executable
RUN chmod +x /custom-entrypoint.sh

# Set the custom script as the entrypoint
ENTRYPOINT ["/custom-entrypoint.sh"]
```

```sh
#!/bin/bash
sed -i 's/max_connections: int = 1/max_connections: int = 100/' snuba/clusters/cluster.py

exec "$@"
```

**Pull request merged**